src/main/java/htsjdk/samtools/cram/build/Sam2CramRecordFactory.java

```diff
@@ -291,36 +271,62 @@ private void addInsertion(final List<ReadFeature> features, final int zeroBasedP
+     * <ul><li>emit nothing for a read base matching corresponding reference base.</li>
+     * <li>emit a {@link Substitution} read feature for each ACTGN-ACTGN mismatch.</li>
+     * <li>emit {@link ReadBase} for a non-ACTGN mismatch. The side effect is the quality score stored twice.</li>
+     * <p>
+     * The rest of the method handles quality score read features added if refSNPs is specified.
+     * <p>
+     * IMPORTANT: reference and read bases are always compared for match/mismatch in upper case due to BAM limitations.
+     *
+     * @param cramRecord           a record to work on
+     * @param features             a list of read features to add to
+     * @param fromPosInRead        a zero based position in the read to start with
+     * @param alignmentStartOffset offset into the reference array
+     * @param nofReadBases         how many read bases to process
+     * @param bases                the read bases array
+     * @param qualityScore         the quality score array
+     */
+    void addSubstitutionsAndMaskedBases(final CramCompressionRecord cramRecord, final List<ReadFeature> features, final int fromPosInRead, final int
             alignmentStartOffset, final int nofReadBases, final byte[] bases, final byte[] qualityScore) {
         int oneBasedPositionInRead;
         final boolean noQS = (qualityScore.length == 0);
 
-        int i;
         boolean qualityAdded;
         byte refBase;
-        for (i = 0; i < nofReadBases; i++) {
+        for (int i = 0; i < nofReadBases; i++) {
             oneBasedPositionInRead = i + fromPosInRead + 1;
             final int referenceCoordinates = cramRecord.alignmentStart + i + alignmentStartOffset - 1;
             qualityAdded = false;
             if (referenceCoordinates >= refBases.length) refBase = 'N';
             else refBase = refBases[referenceCoordinates];
-            refBase = Utils.normalizeBase(refBase);
-
-            if (bases[i + fromPosInRead] != refBase) {
-                final Substitution substitution = new Substitution();
-                substitution.setPosition(oneBasedPositionInRead);
-                substitution.setBase(bases[i + fromPosInRead]);
-                substitution.setReferenceBase(refBase);
 
-                features.add(substitution);
-
-                if (noQS) continue;
+            // explicitly upper case reference base:
+            refBase = SequenceUtil.upperCase(refBase);
+            // explicitly upper case read base:
+            final byte readBase = SequenceUtil.upperCase(bases[i + fromPosInRead]);
+
+            if (readBase != refBase) {
+                final boolean isSubstitution = isACGTN(readBase) && isACGTN(refBase);
+                if (isSubstitution) {
+                    final Substitution substitution = new Substitution();
+                    substitution.setPosition(oneBasedPositionInRead);
+                    substitution.setBase(readBase);
+                    substitution.setReferenceBase(refBase);
```

Comment:
done. I'm not going to replace all calls to default Substitution constructor now to avoid touching too many unrelated classes